imi/src/main/java/org/motechproject/nms/imi/service/impl/CdrFileServiceImpl.java

```diff
@@ -52,36 +72,93 @@
+    private void reportAuditAndThrow(String file, List<String> errors) {
+        for (String error : errors) {
+            LOGGER.error(error);
+            alertService.create(file, "CDR Detail File", error, AlertType.CRITICAL, AlertStatus.NEW, 0, null);
+            fileAuditRecordDataService.create(new FileAuditRecord(FileType.CDR_DETAIL_FILE, file, false, error,
+                    null, null));
+        }
+        throw new InvalidCdrFileException(errors);
+    }
+
+
+    private void sendNotificationRequest(CdrFileProcessedNotification cfpn) {
+        String notificationUrl = settingsFacade.getProperty(CDR_FILE_NOTIFICATION_URL);
+        LOGGER.debug("Sending {} to {}", cfpn, notificationUrl);
+
+        try {
+            HttpClient httpClient = HttpClients.createDefault();
+            HttpPost httpPost = new HttpPost(notificationUrl);
+            ObjectMapper mapper = new ObjectMapper();
+            String requestJson = mapper.writeValueAsString(cfpn);
+            httpPost.setHeader("Content-type", "application/json");
+            httpPost.setEntity(new StringEntity(requestJson));
+            HttpResponse response = httpClient.execute(httpPost);
+            int responseCode = response.getStatusLine().getStatusCode();
+            if (responseCode != HttpStatus.SC_OK) {
+                String error = String.format("Expecting HTTP 200 response from %s but received HTTP %d : %s ",
+                        notificationUrl, responseCode, EntityUtils.toString(response.getEntity()));
+                LOGGER.error(error);
+                alertService.create("cdrFile notification request", cfpn.getFileName(), error,
+                        AlertType.CRITICAL, AlertStatus.NEW, 0, null);
+            }
+        } catch (IOException e) {
+            LOGGER.error("Unable to send cdrFile notification request: {}", e.getMessage());
+            alertService.create("cdrFile notification request", cfpn.getFileName(), e.getMessage(),
+                    AlertType.CRITICAL, AlertStatus.NEW, 0, null);
+        }
+    }
+
+
+    private void reportAuditAndPost(String file, List<String> errors) {
+        for (String error : errors) {
+            LOGGER.error(error);
+            alertService.create(file, "CDR Detail File", error, AlertType.CRITICAL, AlertStatus.NEW, 0, null);
+            fileAuditRecordDataService.create(new FileAuditRecord(FileType.CDR_DETAIL_FILE, file, false, error,
+                    null, null));
+        }
+        sendNotificationRequest(new CdrFileProcessedNotification(
+                FileProcessedStatus.FILE_ERROR_IN_FILE_FORMAT.getValue(),
+                file,
+                StringUtils.join(errors, ",")
+        ));
+    }
+
+
+    // Aggregate detail records for a requestId into single summary records using a (potentially huge) map
+    private void aggregateDetailRecord(CallDetailRecordDto cdr, Map<String, CallSummaryRecordDto> records) {
```

Comment:
I wonder if we will want to do some rough calculations on the memory usage.  A googleing says a java HashMap uses 32 * SIZE + 4 * CAPACITY bytes.  So each CallSummaryRecordDto has an internal map<Integer, Integer>, plus the large map of records.  Additionally the Dto itself will take up some amount of space.

We should certainly put a ticket in to test this in Sprint 4 at least.